File: Core/src/test/java/com/github/ezauton/core/test/simulator/SimulatorTest.java
package com.github.ezauton.core.test.simulator;

import com.github.ezauton.core.action.*;
import com.github.ezauton.core.localization.estimators.TankRobotEncoderEncoderEstimator;
import com.github.ezauton.core.simulation.TimeWarpedSimulation;
import com.github.ezauton.core.simulation.SimulatedTankRobot;
import com.github.ezauton.core.utils.ManualClock;
import org.junit.jupiter.api.Test;

import java.util.concurrent.TimeUnit;
import java.util.concurrent.atomic.AtomicBoolean;
import java.util.concurrent.atomic.AtomicInteger;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertTrue;

public class SimulatorTest
{

    @Test
    public void testSimpleAction()
    {
        AtomicBoolean atomicBoolean = new AtomicBoolean(false);
        TimeWarpedSimulation simulation = new TimeWarpedSimulation();
        simulation.add(new BaseAction(() -> atomicBoolean.set(true)));
        simulation.runSimulation(100, TimeUnit.SECONDS);
        assertTrue(atomicBoolean.get());
    }

    @Test
    public void testDelayedAction()
    {
        AtomicBoolean atomicBoolean = new AtomicBoolean(false);
        TimeWarpedSimulation simulation = new TimeWarpedSimulation();
        DelayedAction delayedAction = new DelayedAction(1, TimeUnit.SECONDS, () -> atomicBoolean.set(true));
        simulation.add(delayedAction);
        simulation.runSimulation(100, TimeUnit.SECONDS);
        assertTrue(atomicBoolean.get());
    }

    @Test
    public void testActionGroup()
    {
        AtomicInteger atomicInteger = new AtomicInteger(0);

        TimeWarpedSimulation simulation = new TimeWarpedSimulation(10);
        ActionGroup actionGroup = new ActionGroup();

        DelayedAction delayedAction = new DelayedAction(1, TimeUnit.SECONDS, () -> atomicInteger.compareAndSet(2, 3));
        delayedAction.onFinish(() -> {});

        DelayedAction delayedAction2 = new DelayedAction(10, TimeUnit.MILLISECONDS, () -> atomicInteger.compareAndSet(0, 1));
        delayedAction2.onFinish(() -> {});

        DelayedAction delayedAction3 = new DelayedAction(500, TimeUnit.MILLISECONDS, () -> atomicInteger.compareAndSet(1, 2));
        delayedAction3.onFinish(() -> {});

        //TODO: Order matters? See github #35
        actionGroup.addParallel(delayedAction3); // second
        actionGroup.with(delayedAction2); // first
        actionGroup.addSequential(delayedAction); // last

        simulation.add(actionGroup);
        simulation.runSimulation(100, TimeUnit.SECONDS);
        assertEquals(3, atomicInteger.get());
    }


    @Test
    public void testStraight()
    {
        ManualClock clock = new ManualClock();
        SimulatedTankRobot robot = new SimulatedTankRobot(1, clock, 14, 0.3, 16);
        TankRobotEncoderEncoderEstimator encoderRotationEstimator = new TankRobotEncoderEncoderEstimator(robot.getLeftDistanceSensor(), robot.getRightDistanceSensor(), robot);
        encoderRotationEstimator.reset();
        for(int i = 0; i < 1000; i++)
        {
            robot.run(1, 1);
            encoderRotationEstimator.update();
            clock.incAndGet();
        }
//        System.out.println("encoderRotationEstimator = " + encoderRotationEstimator.estimateLocation());
    }

    @Test
    public void testTimeout() throws InterruptedException
    {
        AtomicInteger atomicInteger = new AtomicInteger(0);

        TimeWarpedSimulation simulation = new TimeWarpedSimulation(1);
        ActionGroup actionGroup = new ActionGroup();

        PeriodicAction action = new BackgroundAction(20, TimeUnit.MILLISECONDS, atomicInteger::incrementAndGet);

        //TODO: Order matters? See github #35
        actionGroup.addSequential(action);

        simulation.add(actionGroup);
        simulation.runSimulation(1, TimeUnit.SECONDS);

        int actual = atomicInteger.get();
        assertEquals(50, actual, 2);
        Thread.sleep(500); // other threads should have stopped, no more incrementing
//        assertEquals(2, actual, atomicInteger.get());
    }
}

File: Core/src/test/kotlin/com/github/ezauton/core/simulator/SimulatorTest.kt
package com.github.ezauton.core.simulator

import com.github.ezauton.conversion.*
import com.github.ezauton.core.action.*
import com.github.ezauton.core.simulation.SimulatedTankRobot
import com.github.ezauton.core.simulation.parallel
import com.github.ezauton.core.simulation.sequential
import com.github.ezauton.core.utils.RealClock
import kotlinx.coroutines.TimeoutCancellationException
import kotlinx.coroutines.runBlocking
import org.junit.jupiter.api.Assertions.*
import org.junit.jupiter.api.Test
import java.util.concurrent.ExecutionException
import java.util.concurrent.TimeoutException
import java.util.concurrent.atomic.AtomicInteger

class SimulatorTest {

  @Test
  @Throws(TimeoutException::class, ExecutionException::class)
  fun testSimpleAction() = runBlocking {
    var bool = false
    val action = action {
      bool = true
    }

    action.runWithTimeout(1.seconds)

    assertTrue(bool)
  }

  @Test
  @Throws(TimeoutException::class, ExecutionException::class)
  fun testDelayedAction() = runBlocking {

    var bool = false

    val delayedAction = action {
      delay(1.seconds)
      bool = true
    }

    delayedAction.runWithTimeout(2.seconds)

    assertTrue(bool)
  }

  @Test
  @Throws(TimeoutException::class, ExecutionException::class)
  fun testActionGroup() = runBlocking {
    val atomicInteger = AtomicInteger(0)

    val delayedAction = action {
      delay(1.seconds)
      atomicInteger.compareAndSet(2, 3)
      return@action
    }

    val delayedAction2 = action {
      delay(10.ms)
      atomicInteger.compareAndSet(0, 1)
      return@action
    }

    val delayedAction3 = action {
      delay(500.ms)
      atomicInteger.compareAndSet(1, 2)
      return@action
    }


    val action = action {
      parallel(delayedAction3)
      ephemeralScope {
        parallel(delayedAction2)
        sequential(delayedAction)
      }
    }

    action.runWithTimeout(10.seconds)
    assertEquals(3, atomicInteger.get())
  }

  @Test
  fun testStraight() = runBlocking {

    val clock = RealClock
    val robot = SimulatedTankRobot(1.0.meters, clock, 14.0.mpss, 0.3.mps, 16.0.mps)

    for (i in 0..100) {
      robot.run(1.0.mps, 1.0.mps)
      delay(10.ms)
    }

    assertTrue(robot.locationEstimator.estimateLocation().y > 1.m)
    assertTrue(robot.locationEstimator.estimateLocation().x.abs() < 0.01.m)


  }

  @Test
  @Throws(InterruptedException::class)
  fun testTimeout() {
    val atomicInteger = AtomicInteger(0)


    val action = periodicAction(20.ms) {
      atomicInteger.incrementAndGet()
    }

    val actionGroup = action {
      sequential(action)
    }

    assertThrows(TimeoutCancellationException::class.java) {
      runBlocking {
        actionGroup.runWithTimeout(1.seconds)
      }
    }

    runBlocking {
      val actual = atomicInteger.get()
      assertEquals(50f, actual.toFloat(), 2f)
      delay(500.ms) // other threads should have stopped, no more incrementing
      assertEquals(2.0f, actual.toFloat(), atomicInteger.get().toFloat())

    }
  }
}
